core/shared/src/main/scala/fs2/Chunk.scala

```diff
@@ -1635,10 +887,134 @@ object Chunk extends CollectorK[Chunk] with ChunkCompanionPlatform {
       (Chunk.buffer(bldr.result()), cur)
     }
 
-  implicit def fs2EqForChunk[A: Eq]: Eq[Chunk[A]] =
+  /** A FIFO queue of chunks that provides an O(1) size method and provides the ability to
+    * take and drop individual elements while preserving the chunk structure as much as possible.
+    *
+    * This is similar to a queue of individual elements but chunk structure is maintained.
+    */
+  final class Queue[+O] private (val chunks: SQueue[Chunk[O]], val size: Int) extends Chunk[O] {
+    override def iterator: Iterator[O] = chunks.iterator.flatMap(_.iterator)
+
+    override def ++[O2 >: O](that: Chunk[O2]): Chunk[O2] =
+      if (that.isEmpty) this
+      else if (isEmpty) that
+      else new Queue(chunks :+ that, size + that.size)
+
+    /** Prepends a chunk to the start of this chunk queue. */
+    def +:[O2 >: O](c: Chunk[O2]): Queue[O2] = new Queue(c +: chunks, c.size + size)
+
+    /** Appends a chunk to the end of this chunk queue. */
+    def :+[O2 >: O](c: Chunk[O2]): Queue[O2] = new Queue(chunks :+ c, size + c.size)
+
+    def apply(i: Int): O = {
+      if (i < 0 || i >= size) throw new IndexOutOfBoundsException()
+      def go(chunks: SQueue[Chunk[O]], offset: Int): O = {
+        val head = chunks.head
+        if (offset < head.size) head(offset)
+        else go(chunks.tail, offset - head.size)
+      }
+      go(chunks, i)
+    }
+
+    override def take(n: Int): Queue[O] =
```

Comment:
An interesting fact

According to the results of this benchmark
```

package fs2
package benchmark

import scala.util.Random
import cats.effect.IO
import cats.effect.unsafe.implicits.global
import org.openjdk.jmh.annotations.{Benchmark, Param, Scope, Setup, State}

@State(Scope.Thread)
class ChunkQueueBenchmark {
  @Param(Array("10", "100", "10000"))
  var chunkSize: Int = _

  var chunk: Chunk.Queue[Int] = _
  @Setup
  def setup(): Unit = {
    val seq = Seq.fill(100_000/chunkSize)(Chunk.seq(Seq.fill(chunkSize)(Random.nextInt())))
    chunk = Chunk.Queue(seq: _*)
  }

  @Benchmark
  def drop(): Unit =
    chunk.drop(chunkSize - 1)

  @Benchmark
  def take(): Unit =
    chunk.take(chunkSize - 1)
}
```

`drop` doesn't degrade when increasing SIZE of chunks, while `take` does.
```
[info] ChunkQueueBenchmark.drop           10  thrpt       21521956,531          ops/s
[info] ChunkQueueBenchmark.drop          100  thrpt       21206792,447          ops/s
[info] ChunkQueueBenchmark.drop        10000  thrpt       19296012,270          ops/s

[info] ChunkQueueBenchmark.take           10  thrpt       13121396,574          ops/s
[info] ChunkQueueBenchmark.take          100  thrpt        2185148,679          ops/s
[info] ChunkQueueBenchmark.take        10000  thrpt          21255,437          ops/s
```

Reply:
How about `chunk.drop(1)`?

Reply:
oh yes, I messed up the benchmark, I used `chunkSize - 1` instead of `100_000 - 1` thanks! :flushed:

Reply:
Previously I thought, that appending to the end of immutable queue is linear. In `take(n)` elements are appended, thus I thought that it would be n^2 to the amount of chunks. However, I was wrong. TIL, that appending to both the end and the head of the immutable queue in scala is constant, as is head(pop first) and dequeue(pop last).

Reply:
`scala.collection.immutable.Queue` doesn't get nearly enough credit :)